presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java

```diff
@@ -51,20 +51,20 @@
 
 import static com.facebook.presto.cost.ComparisonStatsCalculator.estimateExpressionToExpressionComparison;
```

Comment:
I'm not a big fan of this change. Can we work out a proper semantics (it really should be doable) for `subtractColumnStats` instead of simply removing it and having non-obvious code duplicated? Maybe the method had semi-defined semantics but after this change we have an implicit semantics for rather complex logic.

The advantage of dedicated `subtractColumnStats` would be also that it semantics can be properly described in Javadoc. For stats code this is not alwyas obvious from code. I also worry that if there comes a need, there will be 3rd and 4th implementation of column stats subtracting (as people will think they need special handling and there isn't stats toolset), which is not desirable.

Reply:
Let me explain why did i decided to inline it. That method was used only for `OR` and for `NOT` estimates. The `OR` estimation involves both, summing stats and than subtracting `AND` stats. It seems that it would be much easier to follow that code if no higher level stats method were involved. In other words when doing both, `summing` and `substracting` inline. The code seems to be more transparent, and easier to follow

e.g.:

- setting low-high only taking into account both disjunts separately: https://github.com/prestodb/presto/pull/11973/files#diff-2c8e7079274d1e50a57e44770854b2c5R324
- setting avg size to the input avg size: https://github.com/prestodb/presto/pull/11973/files#diff-2c8e7079274d1e50a57e44770854b2c5R322
- setting ndvs in a more explicit manner that is easier to follow: https://github.com/prestodb/presto/pull/11973/files#diff-2c8e7079274d1e50a57e44770854b2c5R328

Now since the code for `OR` got inlined i wasn't sure that we need a separate "abstract" method that effective is used just to estimate the negation. For me it is also easier to read. As you explicitly assuming that what you are doing - is negating stats. Not subtracting them. 

Thoughts?

Reply:
IMHO there is no such concept as "stats subtraction". You can subtract numbers, but "10 NDV - 15 NDV" does not make sense, in general.
For `NOT` we want to take "complement" which is a special case of "subtraction" (and still, may be different than just numerical subtraction of each stat property). Do we have other places where exactly such a "complement" is being calculated? Do we do this for outer joins?

Reply:
> IMHO there is no such concept as "stats subtraction".

There is a concept of multiset (e.g: rows) subtraction and those sets are described by stats.

> You can subtract numbers, but "10 NDV - 15 NDV" does not make sense, in general.

It does with additional assumptions. We do something similar for join when we assume that NDVs match.

Anyaway, I propsed subtract implementation in comment.